File: apple/src/StreamInformation.m
/*
 * Copyright (c) 2018-2022 Taner Sener
 *
 * This file is part of FFmpegKit.
 *
 * FFmpegKit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with FFmpegKit.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "StreamInformation.h"

NSString* const StreamKeyIndex = @"index";
NSString* const StreamKeyType = @"codec_type";
NSString* const StreamKeyCodec = @"codec_name";
NSString* const StreamKeyCodecLong = @"codec_long_name";
NSString* const StreamKeyFormat = @"pix_fmt";
NSString* const StreamKeyWidth = @"width";
NSString* const StreamKeyHeight = @"height";
NSString* const StreamKeyBitRate = @"bit_rate";
NSString* const StreamKeySampleRate = @"sample_rate";
NSString* const StreamKeySampleFormat = @"sample_fmt";
NSString* const StreamKeyChannelLayout = @"channel_layout";
NSString* const StreamKeySampleAspectRatio = @"sample_aspect_ratio";
NSString* const StreamKeyDisplayAspectRatio = @"display_aspect_ratio";
NSString* const StreamKeyAverageFrameRate = @"avg_frame_rate";
NSString* const StreamKeyRealFrameRate = @"r_frame_rate";
NSString* const StreamKeyTimeBase = @"time_base";
NSString* const StreamKeyCodecTimeBase = @"codec_time_base";
NSString* const StreamKeyTags = @"tags";

@implementation StreamInformation {

    /**
     * Stores all properties.
     */
    NSDictionary *dictionary;

}

- (instancetype)init:(NSDictionary*)streamDictionary {
    self = [super init];
    if (self) {
        dictionary = streamDictionary;
    }

    return self;
}

- (NSNumber*)getIndex {
    return [self getNumberProperty:StreamKeyIndex];
}

- (NSString*)getType {
    return [self getStringProperty:StreamKeyType];
}

- (NSString*)getCodec {
    return [self getStringProperty:StreamKeyCodec];
}

- (NSString*)getCodecLong {
    return [self getStringProperty:StreamKeyCodecLong];
}

- (NSString*)getFormat {
    return [self getStringProperty:StreamKeyFormat];
}

- (NSNumber*)getWidth {
    return [self getNumberProperty:StreamKeyWidth];
}

- (NSNumber*)getHeight {
    return [self getNumberProperty:StreamKeyHeight];
}

- (NSString*)getBitrate {
    return [self getStringProperty:StreamKeyBitRate];
}

- (NSString*)getSampleRate {
    return [self getStringProperty:StreamKeySampleRate];
}

- (NSString*)getSampleFormat {
    return [self getStringProperty:StreamKeySampleFormat];
}

- (NSString*)getChannelLayout {
    return [self getStringProperty:StreamKeyChannelLayout];
}

- (NSString*)getSampleAspectRatio {
    return [self getStringProperty:StreamKeySampleAspectRatio];
}

- (NSString*)getDisplayAspectRatio {
    return [self getStringProperty:StreamKeyDisplayAspectRatio];
}

- (NSString*)getAverageFrameRate {
    return [self getStringProperty:StreamKeyAverageFrameRate];
}

- (NSString*)getRealFrameRate {
    return [self getStringProperty:StreamKeyRealFrameRate];
}

- (NSString*)getTimeBase {
    return [self getStringProperty:StreamKeyTimeBase];
}

- (NSString*)getCodecTimeBase {
    return [self getStringProperty:StreamKeyCodecTimeBase];
}

- (NSDictionary*)getTags {
    return [self getProperty:StreamKeyTags];
}

- (NSString*)getStringProperty:(NSString*)key {
    NSDictionary* allProperties = [self getAllProperties];
    if (allProperties == nil) {
        return nil;
    }

    return allProperties[key];
}

- (NSNumber*)getNumberProperty:(NSString*)key {
    NSDictionary* allProperties = [self getAllProperties];
    if (allProperties == nil) {
        return nil;
    }

    return allProperties[key];
}

- (id)getProperty:(NSString*)key {
    NSDictionary* allProperties = [self getAllProperties];
    if (allProperties == nil) {
        return nil;
    }

    return allProperties[key];
}

- (NSDictionary*)getAllProperties {
    return dictionary;
}

@end

File: linux/src/StreamInformation.cpp
/*
 * Copyright (c) 2022 Taner Sener
 *
 * This file is part of FFmpegKit.
 *
 * FFmpegKit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with FFmpegKit.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "StreamInformation.h"

ffmpegkit::StreamInformation::StreamInformation(std::shared_ptr<rapidjson::Value> streamInformationValue) : _streamInformationValue{streamInformationValue} {
}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getIndex() {
    return getNumberProperty(KeyIndex);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getType() {
    return getStringProperty(KeyType);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getCodec() {
    return getStringProperty(KeyCodec);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getCodecLong() {
    return getStringProperty(KeyCodecLong);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getFormat() {
    return getStringProperty(KeyFormat);
}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getWidth() {
    return getNumberProperty(KeyWidth);
}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getHeight() {
    return getNumberProperty(KeyHeight);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getBitrate() {
    return getStringProperty(KeyBitRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getSampleRate() {
    return getStringProperty(KeySampleRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getSampleFormat() {
    return getStringProperty(KeySampleFormat);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getChannelLayout() {
    return getStringProperty(KeyChannelLayout);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getSampleAspectRatio() {
    return getStringProperty(KeySampleAspectRatio);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getDisplayAspectRatio() {
    return getStringProperty(KeyDisplayAspectRatio);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getAverageFrameRate() {
    return getStringProperty(KeyAverageFrameRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getRealFrameRate() {
    return getStringProperty(KeyRealFrameRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getTimeBase() {
    return getStringProperty(KeyTimeBase);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getCodecTimeBase() {
    return getStringProperty(KeyCodecTimeBase);
}

std::shared_ptr<rapidjson::Value> ffmpegkit::StreamInformation::getTags() {
    return getProperty(KeyTags);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getStringProperty(const char* key) {
    if (_streamInformationValue->HasMember(key)) {
        return std::make_shared<std::string>((*_streamInformationValue)[key].GetString());
    } else {
        return nullptr;
    }
}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getNumberProperty(const char* key) {
    if (_streamInformationValue->HasMember(key)) {
        return std::make_shared<int64_t>((*_streamInformationValue)[key].GetInt64());
    } else {
        return nullptr;
    }
}

std::shared_ptr<rapidjson::Value> ffmpegkit::StreamInformation::getProperty(const char* key) {
    if (_streamInformationValue->HasMember(key)) {
        auto value = std::make_shared<rapidjson::Value>();
        *value = (*_streamInformationValue)[key];
        return value;
    } else {
        return nullptr;
    }
}

std::shared_ptr<rapidjson::Value> ffmpegkit::StreamInformation::getAllProperties() {
    if (_streamInformationValue != nullptr) {
        auto all = std::make_shared<rapidjson::Value>();
        *all =  (*_streamInformationValue);
        return all;
    } else {
        return nullptr;
    }
}
